flow-nodes/api-builder-plugin-format-date/src/actions.js

```diff
@@ -21,24 +21,60 @@ const moment = require('moment');
  *	 does not define "next", the first defined output).
  */
 function formatDate(params) {
-	const { date, format } = params;
+	let { date, format, offset } = params;
 
-	if (!date) {
-		throw new Error('Missing required parameter: date');
-	}
-	if (!format) {
-		throw new Error('Missing required parameter: format');
+  let formattedDate;
+  let dateString = (!date) ? date : date.toString(); // Convert any date input such as "new Date('2020-01-08T13:24:40+02:00')" to a string.
+  let dateWithNoOrdinal = (!date) ? date : dateString.replace(/(\d+)(st|nd|rd|th)/, "$1");
+  let dateMoment = moment(dateWithNoOrdinal);
+  let isDate = dateMoment.isValid();
+  let formatWithNoOrdinal = (!format) ? format : format.replace(/Do/g, 'D');
+
+  if (!date && !format) {
```

Comment:
Are we not to check the validity of the format entered by the user as well?
This is what I had for an invalid format and is it the expectation?
`{ date: ’2020-08-05’, format: 'clearly-not-a-format' } => 'c8/4/20202pmr8/4/20202020-not-pm-for0pmt'`

Also, should we not allow the user to be able to format the current datetime by only entering the desired format and not passing in anything for the date? As in a case when only the format parameter is passed should format the current datetime. What's your opinion on this?

Reply:
It's not for us to determine if it's a valid format - moment was okay with "clearly-not-a-format", so it's fine.  we only need to test the exceptional cases when moment returns an error.

For the "format only" parameter, yes, that should format the current datetime in UTC time, using whatever format they passed in.

Reply:
@Dankyi  clearly-not-a-format is returning something a bit weird, but technically it is a valid format. If users wanted to return a very custom date format then they could make use of any number of characters, spacers and/or seperators in their format.

A good example is the following format:

`'[It\'s the] Do [ day of the month] [of] MMMM.'`

`It's the 5th day of the month of November.`

Reply:
I think it be simplified even further actually:
```
function formatDate(params) {
	const { date, format, offset } = params;

        // Defaults to now
	const momentObj = moment(date);

	if (offset) {
		momentObj.utcOffset(offset);
	} else {
		momentObj.utc();
	}
       
       // Defaults to ISO-8601 (YYYY-MM-DDTHH:mm:ssZ)
        const formattedDate = momentObj.format(format);
	if (formattedDate === 'Invalid date') {
		throw new Error(`Invalid date: ${date}`);
	}
	return formattedDate;
}
```

Although i would still recommend making the defaults clearer

Reply:
But since we don't validate selectors, we'd need a bit of extra parameter validation here - i.e. checking the type of offset if we're only supporting string offsets. (although at this point i'd think we'd be fine to support number and string tbh and save some code/tests)

Reply:
@alasdairhurst, you're not wrong, but why be so specifically careful with `offset`?  When any parameter is passed via a selector, it could be literally anything.  We don't normally do this kind of type checking.

Reply:
not specifically offset - was just an example. we should be validating all our inputs since the schema doesn't guarantee any control